trin-state/src/utils.rs

```diff
@@ -18,6 +23,69 @@ const MID: [u8; 77] = [
     49, 57, 57, 54, 56,
 ];
 
+pub fn state_network_derive_content_id(key: &String) -> U256 {
+    // SSZ decode ContentKey
+    let decoded_key = match StateContentKey::from_bytes(&key.as_bytes()) {
+        Ok(k) => k,
+        Err(_) => panic!("Cannot decode key")
+    };
+
+    match decoded_key {
+        StateContentKey::AccountTrieNode(AccountTrieNode { path, node_hash, .. }) => {
+            let mut input = String::from_utf8(path).unwrap();
+            input.push_str(&node_hash.to_string());
+            sha256(&input)
+        },
+        StateContentKey::ContractStorageTrieNode(ContractStorageTrieNode { address, path, node_hash, .. }) => {
+            // sha256(address, path, node_hash)
+            let mut address = address.clone();
+            let mut path = path.clone();
+            address.append(&mut path);
+
+            let mut input = String::from_utf8(address).unwrap();
+            input.push_str(&node_hash.to_string());
+            sha256(&input)
+        },
+        StateContentKey::AccountTrieProof(AccountTrieProof { address, .. }) => {
+            sha256(&String::from_utf8(address).unwrap())
+        },
+        StateContentKey::ContractStorageTrieProof(ContractStorageTrieProof { address, slot, .. }) => {
+            let address_hash = keccak(&String::from_utf8(address).unwrap());
+            let slot_hash = keccak(&slot.to_string());
+            let final_hash: U256 = address_hash + slot_hash;
+
+            let base: U256 = U256::from(2);
+            let exp: U256 = U256::from(256);
+
+            return final_hash % base.pow(exp);
+        },
+        StateContentKey::ContractBytecode(ContractBytecode { address, code_hash }) => {
+            let mut input = String::from_utf8(address).unwrap();
+            input.push_str(&code_hash.to_string());
+            sha256(&input)
+        },
+    }
+}
+
+fn vec_to_array<T, const N: usize>(v: Vec<T>) -> [T; N] {
+    v.try_into()
+      .unwrap_or_else(|v: Vec<T>| panic!("Expected a Vec of length {} but it was {}", N, v.len()))
+}
+
+fn sha256(key: &String) -> U256 {
+    let mut hasher = Sha3_256::new();
+    hasher.update(key);
+    let result = hasher.finalize().to_vec();
+    U256::from(vec_to_array(result))
+}
+
+fn keccak(key: &String) -> U256 {
+    let mut hasher = Keccak256::new();
```

Comment:
Ah, interesting, I may have missed this convo. Are we really switching to sha3_256 proper? There's a lot of room for confusion, for it to live alongside keccak256. Wondering if the spec was just a little loose in saying sha, when it meant keccak.

Reply:
The crate I am using has both sha3_256 and Keccak256. It notes that Keccak256 is the NIST submission without padding changes. So I'm guessing sha3_256 is the actual "controversial" standard? The specs for state and history specify both "sha256" and just "keccak" so my guess is they mean the SHA-2 family for "sha256" and "keccak" as the "unproper". @pipermerriam